alws/crud/test.py

```diff
@@ -266,15 +266,33 @@ async def restart_build_tests(db: AsyncSession, build_id: int):
     # which only restarts tests for those builds that already
     # had passed the tests
     async with db.begin():
-        build_task_ids = await db.execute(
-            select(models.BuildTask.id).where(
-                models.BuildTask.build_id == build_id
-            )
+        query = (
+            select(models.BuildTask)
+            .options(joinedload(models.BuildTask.test_tasks))
+            .where(models.BuildTask.build_id == build_id)
         )
-        build_task_ids = build_task_ids.scalars().all()
+
+        build_tasks = await db.execute(query)
+        build_tasks = build_tasks.scalars().unique().all()
         test_log_repository = await __get_log_repository(db, build_id)
-    for build_task_id in build_task_ids:
-        await create_test_tasks(db, build_task_id, test_log_repository.id)
+    for build_task in build_tasks:
+        if not build_task.test_tasks:
+            continue
+
+        last_reviosion = build_task.test_tasks[-1].revision
+        failed = False
+        for test_task in build_task.test_tasks:
+            if test_task.revision != last_reviosion:
+                continue
+
+            if test_task.status == TestTaskStatus.FAILED:
+                failed = True
```

Comment:
Maybe we should break out of loop after failed=True? I don`t see why we need to continue